### PR TITLE
build: full margin optimizations

### PR DESCRIPTION
**Pull request by antoncoding**
## Updates
* Use calldata directly in `BaseEngine`, safe 180 gas per action (copy data to memory) 
* remove unnecessary storage reads for events
* remove duplicated 0 amount check in `getMinCollateral`. Saves few hundred ~ 2000 gas

### CODE REVIEW

File: src/core/engines/cross-margin/CrossMarginMath.sol [hunk 1]
```diff
@@ -69,7 +70,7 @@ library CrossMarginMath {
             CrossMarginDetail memory detail = details[i];
 
             if (detail.callWeights.length != 0 || detail.putWeights.length != 0) {
-                (int256 cashCollateralNeeded, int256 underlyingNeeded) = getMinCollateral(detail);
+                (int256 cashCollateralNeeded, int256 underlyingNeeded) = _getMinCollateralWithoutVerification(detail);
```
Comment (dsshap) [inline src/core/engines/cross-margin/CrossMarginMath.sol:73]: We still want verification

Comment (dsshap) [reply src/core/engines/cross-margin/CrossMarginMath.sol:73]: I'll update this after merging in

Comment (antoncoding) [reply src/core/engines/cross-margin/CrossMarginMath.sol:73]: the details you get from `_getAccountDetails` should not contain any 0 amounts

**Review by dsshap — APPROVED** [review]

File: src/core/engines/cross-margin/CrossMarginMath.sol (hunk 1 above)
Comment (antoncoding) [reply src/core/engines/cross-margin/CrossMarginMath.sol:73]: oh maybe we skip the 0 amount check but still need to length check?

**Pull request merged**